File: src/GnOuGo.Flow.Core/Runtime/Executors/EmitExecutor.cs
using System.Text.Json.Nodes;
using GnOuGo.Flow.Core.Expressions;
using GnOuGo.Flow.Core.Models;

namespace GnOuGo.Flow.Core.Runtime.Executors;

/// <summary>
/// Emits a thinking / progress / info / response message to the telemetry stream.
/// Allows workflows to push real-time feedback to the user interface.
///
/// Input:
///   - message (string, required) : The message to emit.
///   - level   (string, optional) : One of "thinking", "info", "progress", "response". Default: "thinking".
///
/// Output:
///   { message: "...", level: "..." }
/// </summary>
public sealed class EmitExecutor : IStepExecutor
{
    public string StepType => "emit";

    private static readonly HashSet<string> ValidLevels = new(StringComparer.OrdinalIgnoreCase)
    {
        "thinking", "info", "progress", "response"
    };

    public IReadOnlyList<StepExceptionDoc>? DocumentedExceptions => new StepExceptionDoc[]
    {
        new(ErrorCodes.InputValidation, false, "The input object is malformed or the 'message' field is missing.")
    };

    public string DslSnippet => """
        ### emit — Emit a thinking / progress message
        Sends a real-time message to the user interface. Useful for showing progress,
        thinking steps, or partial responses during long-running workflows.
        Levels: `thinking` (default, subtle animated), `info` (blue), `progress` (green), `response` (highlighted, monospace).
        ```yaml
        - id: thinking_step
          type: emit
          input:
            message: "Analyzing the data..."
            level: thinking

        - id: progress_step
          type: emit
          input:
            message: "Processed 3 of 5 items"
            level: progress

        - id: early_response
          type: emit
          input:
            message: "Here is a preliminary answer: ..."
            level: response
        ```
        Output: `{ message: "...", level: "..." }`
        """;

    public Task<JsonNode?> ExecuteAsync(StepExecutionContext ctx, CancellationToken ct)
    {
        var input = ctx.Engine.GetResolvedInput(ctx) as JsonObject
            ?? throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "emit input must be object");

        var message = input["message"]?.GetValue<string>()
            ?? throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "emit requires 'message'");

        var level = input["level"]?.GetValue<string>()?.ToLowerInvariant() ?? "thinking";
        if (!ValidLevels.Contains(level))
            level = "thinking";

        // Fire a telemetry event so StreamingWorkflowTelemetry can push it to the browser
        ctx.AddTelemetryEvent("gnougo-flow.step.thinking", new[]
        {
            new KeyValuePair<string, object?>("gnougo-flow.thinking.message", message),
            new KeyValuePair<string, object?>("gnougo-flow.thinking.level", level)
        });

        var result = new JsonObject
        {
            ["message"] = message,
            ["level"] = level
        };

        return Task.FromResult<JsonNode?>(result);
    }
}

File: clients/python/gnougo-flow-core/src/gnougo_flow_core/runtime_steps/emit_executor.py
from __future__ import annotations

from gnougo_flow_core.runtime import *  # noqa: F401,F403


class EmitExecutor:
    step_type = "emit"
    step_description = "Emit progress/thinking messages to telemetry/UI."
    dsl_snippet = """
### emit - Send a progress/thinking message
```yaml
- id: thinking
  type: emit
  input:
    message: "Planning next actions"
    level: thinking
```
Output: `{ message, level }`.
"""
    documented_exceptions = [
        (ErrorCodes.INPUT_VALIDATION, False, "The input object is malformed or the 'message' field is missing."),
    ]

    async def execute_async(self, ctx: StepExecutionContext) -> Any:
        input_obj = ctx.engine.get_resolved_input(ctx)
        if not isinstance(input_obj, dict):
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "emit input must be object")
        message = input_obj.get("message")
        if not isinstance(message, str):
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "emit requires 'message'")
        level = str(input_obj.get("level", "thinking")).lower()
        if level not in {"thinking", "info", "progress", "response"}:
            level = "thinking"
        if ctx.telemetry_span:
            ctx.telemetry_span.add_event(
                "gnougo-flow.step.thinking",
                [("gnougo-flow.thinking.message", message), ("gnougo-flow.thinking.level", level)],
            )
        return {"message": message, "level": level}
